### PR TITLE
Update QuantumCircuit properties

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
Changes the way the `QuantumCircuit` properties are computed.

### Details and comments
I have made the following modifications that make the properties more self-consistent and, inline with the properties of the corresponding DAG (where possible):

1) The circuit depth takes into account conditionals and which `clbits` measurements use.  This is because measurements cannot write to the same clbit at once, and a conditional gate cannot be done before measurements before it.  As a result, the depth of the circuit necessarily requires including the clbits and thus they are part of the circuits graph structure.  This motivates (2) and (3) as well.

2) Circuit `width` takes into account `clbits`  This is because the definition of the `width` is defined to be the number of parallel gates that can be run.  It is possible to do gates on the `clbits` as well as the `qubits` in parallel.  If there are no classical bits, then the definition becomes the standard QC one.

3) The number of tensor factors has been split into two functions (1) `connected_components`  that computes the number of connected components over the entire circuit, qubits and cbits. (2) `num_unitary_factors` that computes the number of tensor factors in the unitary (quantum) part of the circuit alone.  `num_tensor_factors` is still a circuit method for backward compatibility, but just calls `num_unitary_factors`.

4) The `size` of the circuit measures gates and measurements only (no barriers or snapshots).  This makes sense as the circuit size is supposed to measure the number of gates, and barriers and snapshots are transpiler and simulator directives, respectively.

### CODE REVIEW

**Comment by nonhermitian** [comment]
I forgot to get the depth working with multiple registers.

**Comment by nonhermitian** [comment]
Ok, now it works.

File: qiskit/circuit/quantumcircuit.py [hunk 1]
```diff
@@ -325,39 +325,193 @@ def draw(self, scale=0.7, filename=None, style=None, output='text',
...
+        # Conditional gates act on all cbits in the register
+        # they are conditioned on.
+        # We do not consider barriers or snapshots as
+        # They are transpiler and simulator directives.
+        # The max stack height is the circuit depth.
+        for op in self.data:
+            if op.name not in ['barrier', 'snapshot']:
+                levels = []
+                reg_ints = []
+                for ind, reg in enumerate(op.qargs+op.cargs):
```
Comment (ajavadia) [inline qiskit/circuit/quantumcircuit.py]: After #1816 the ops no longer have .qargs and .cargs. Instead, when you iterate over `circuit.data` you get tuples of `(op, qargs, cargs)`. So this PR needs to be updated.

Comment (nonhermitian) [reply qiskit/circuit/quantumcircuit.py]: resolved

**Pull request merged**